src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs

```diff
@@ -4720,4 +4799,1022 @@ public class C
+            """;
+        var comp1 = CreateCompilation(source1);
+        comp1.VerifyDiagnostics();
+        var comp1Ref = comp1.ToMetadataReference();
+
+        var source2 = """
+            C.X((ref int p) => System.Console.Write("1"));
+            C.Y((in int p) => System.Console.Write("2"));
+
+            C.X((in int p) => System.Console.Write("3"));
+            C.Y((ref int p) => System.Console.Write("4"));
+            """;
+
+        CreateCompilation(source2, new[] { comp1Ref }, parseOptions: TestOptions.Regular11).VerifyDiagnostics(
+            // (4,5): error CS1661: Cannot convert lambda expression to type 'X' because the parameter types do not match the delegate parameter types
+            // C.X((in int p) => System.Console.Write("3"));
+            Diagnostic(ErrorCode.ERR_CantConvAnonMethParams, @$"(in int p) => System.Console.Write(""3"")").WithArguments("lambda expression", "X").WithLocation(4, 5),
+            // (4,13): error CS1676: Parameter 1 must be declared with the 'ref' keyword
+            // C.X((in int p) => System.Console.Write("3"));
+            Diagnostic(ErrorCode.ERR_BadParamRef, "p").WithArguments("1", "ref").WithLocation(4, 13),
+            // (5,5): error CS1661: Cannot convert lambda expression to type 'Y' because the parameter types do not match the delegate parameter types
+            // C.Y((ref int p) => System.Console.Write("4"));
+            Diagnostic(ErrorCode.ERR_CantConvAnonMethParams, @"(ref int p) => System.Console.Write(""4"")").WithArguments("lambda expression", "Y").WithLocation(5, 5),
+            // (5,14): error CS1676: Parameter 1 must be declared with the 'in' keyword
+            // C.Y((ref int p) => System.Console.Write("4"));
+            Diagnostic(ErrorCode.ERR_BadParamRef, "p").WithArguments("1", "in").WithLocation(5, 14));
+
+        var expectedDiagnostics = new[]
+        {
+            // (4,5): warning CS9510: Reference kind modifier of parameter 'in int p' doesn't match the corresponding parameter 'ref int p' in target.
+            // C.X((in int p) => System.Console.Write("3"));
+            Diagnostic(ErrorCode.WRN_TargetDifferentRefness, @"(in int p) => System.Console.Write(""3"")").WithArguments("in int p", "ref int p").WithLocation(4, 5),
+            // (5,5): warning CS9510: Reference kind modifier of parameter 'ref int p' doesn't match the corresponding parameter 'in int p' in target.
+            // C.Y((ref int p) => System.Console.Write("4"));
+            Diagnostic(ErrorCode.WRN_TargetDifferentRefness, @"(ref int p) => System.Console.Write(""4"")").WithArguments("ref int p", "in int p").WithLocation(5, 5)
+        };
+
+        CompileAndVerify(source2, new[] { comp1Ref }, expectedOutput: "X1Y2X3Y4", parseOptions: TestOptions.RegularNext).VerifyDiagnostics(expectedDiagnostics);
+        CompileAndVerify(source2, new[] { comp1Ref }, expectedOutput: "X1Y2X3Y4").VerifyDiagnostics(expectedDiagnostics);
+    }
```

Comment:
>} [](http://example.com/codeflow?start=4&length=1)

We should test that warnings are reported for complex conversions, like tuple expression conversions, etc.